x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java

```diff
@@ -241,6 +255,93 @@ default AuthorizationInfo getAuthenticatedUserAuthorizationInfo() {
         }
     }
 
+    record PrivilegesToCheck(
+        String[] cluster,
+        RoleDescriptor.IndicesPrivileges[] index,
+        RoleDescriptor.ApplicationResourcePrivileges[] application
+    ) {
+        public static PrivilegesToCheck readFrom(StreamInput in) throws IOException {
+            return new PrivilegesToCheck(
+                in.readOptionalStringArray(),
+                in.readOptionalArray(RoleDescriptor.IndicesPrivileges::new, RoleDescriptor.IndicesPrivileges[]::new),
+                in.readOptionalArray(RoleDescriptor.ApplicationResourcePrivileges::new, RoleDescriptor.ApplicationResourcePrivileges[]::new)
+            );
+        }
+
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeOptionalStringArray(cluster);
+            out.writeOptionalArray(RoleDescriptor.IndicesPrivileges::write, index);
+            out.writeOptionalArray(RoleDescriptor.ApplicationResourcePrivileges::write, application);
+        }
+
+        public ActionRequestValidationException validate(ActionRequestValidationException validationException) {
+            if (cluster == null) {
+                validationException = addValidationError("clusterPrivileges must not be null", validationException);
+            }
+            if (index == null) {
+                validationException = addValidationError("indexPrivileges must not be null", validationException);
+            }
+            if (application == null) {
+                validationException = addValidationError("applicationPrivileges must not be null", validationException);
+            } else {
```

Comment:
The `null` validation is in place because theoretically (but not in practice) the request for the existing has privilege endpoint can be built without these fields being set.
Do you want me to change the request construction for the existing has privilege endpoint?

Reply:
No. Let's keep it this way. As metioned in another comment, I wasn't aware HasPrivilegesRequest is used in many other places. So I don't think this is the PR to change that.